### PR TITLE
Update --labels with additional options.

### LINKED ISSUE #24: Update --labels switch with option to show real-time pass/fail results in console runner
@dybs commented on [Fri Aug 12 2016](https://github.com/nunit/nunit/issues/1738)

When running a large number of test cases, it would be nice to see the pass/fail status of each case in the console as it completes.  Using --labels=All, I can see which tests have run, but the results are not available until all cases have completed.  Seeing which cases fail during a long scenario would allow me to investigate the failures while the remaining tests continue to run.  It also has the advantage of knowing which cases passed/failed if the test scenario happens to crash and doesn't get to write the result file.

Per the discussion in #1735 some ideas are to modify --labels with the following options:
- Results -> same behavior as =All, but includes a Pass/Fail/Error status
- Before -> shows the test case before it executes, and appends the status after the test completes.  Similar to On.
- After -> shows the test case and status together after the test case completes.

---

@CharliePoole commented on [Fri Aug 12 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239549055)

I like this idea, but I would modify/simplify it by preference:
- Off No change
- On No change
- Before / All shows test label when the test starts executing.
- After Shows all tests after they have completed, including status.

Note that All is described as it is supposed to work already. There's currently a bug in that it runs at the end of a test. We should fix that. We should keep All so as not to break existing usage.

The introduction of After means that we have to treat it as also meaning On. That is, if a test produces output, there should be a label before that output and another one after it, together with the test status..

The change should be made in both the console runner and nunitlite. Once we split the repos, this issue will need to be duplicated.

The fix is _relatively_ easy. Depending on who takes it, I can mentor them as needed.

---

@constructor-igor commented on [Sun Aug 14 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239676310)

Hi @CharliePoole,

Probably, I can start to work on the issue.
I checked all existing options of "--labels" parameter. Please, could you help me to understand suggested changes? I added sample with "--lablels=All". As I understand, the issue suggests to add "status" of each test?

image

thank you,
Igor.

---

@CharliePoole commented on [Sun Aug 14 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239684890)

Hi @constructor-igor - I've been meaning to poke you to see what you are up to. :-)

As you probably figured out, all the code is in Console's TestEventHandler and in NUnitLite's TestRunner. There are tests for TestEventHandler but not for TestRunner. :-(

I would first off fix the problem with All, which is supposed to be issued at the start of each test, not at the end. I think I broke this when I was fixing something else.

Once All is correct, you've got Before, since it's intended to work the same way.

After is the new thing, which is supposed to show status of the test. For that, I would use a combination of the `result` and `label` attributes to display a set of values like...
- Passed
- Failed
- Error
- Invalid
- Ignored
- Explicit
- Skipped
- Inconclusive
  In NUnitLite, you'll have the ResultState available to use for equivalent logic.

---

@constructor-igor commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239782084)

I am going to start from "All".
But, right now, I don't understand expected behavior of "All".
So, when I'll investigate code I'll contact with you.

---

@CharliePoole commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239783383)

You might try this sequence - a bit different order from my first suggestion:
- Change "All" in existing code everywhere to "After" because All is now incorrectly running after the test.
- Add status info to After
- Add "Before" to run before the test
- Add back "All" as a synyonym for Before

---

@constructor-igor commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239785294)

I try to build nunit by "build.cmd":
image

VS2013:
image

---

@CharliePoole commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239790081)

Are you using the latest master? Can you build in the IDE?

---

@constructor-igor commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239796767)

I cloned today and tried to build via build.cmd and IDE VS2013: same errors.

---

@CharliePoole commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239798656)

Using VS 2015?

---

@rprouse commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239803785)

I just got latest from master and ran `.\build.cmd` and it builds fine for me. Maybe do a full clean including deleting your packages folder and rebuild?

---

@constructor-igor commented on [Mon Aug 15 2016](https://github.com/nunit/nunit/issues/1738#issuecomment-239894381)

I could build nunit solution on other computer.
I tried to build on my computer again (downloaded master to new folder) and  see the error.

I investigate the issue and found, probably, same issues 
- https://github.com/nunit/nunit3-vs-adapter/issues/194
- https://github.com/nunit/docs/issues/44

I continue my investigation.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #24 

@constructor-igor spent a lot of time on this issue, but the branch was never merged. As part of cleaning up my old work in the framework, I decided to recreate the branch and try to get it finished.

Since the issue had a lot of comments over time, some conflicting, I'll summarize how the new labels values work in this PR. I use "normal" output to mean output through `Console.Out` or `TestContext.Out` and "immediate" for output through `Console.Error`, `TestContext.Progress` or `TestContext.Error`.

**--labels:Off** 
No labeling is used. Both normal and immediate output appear in the order produced - i.e. immediate first.

**--labels:On**
A label appears before each sequence of output lines from the same test. Since tests may be run in parallel, output from different tests may be intermixed.

**--labels:All**
A label appears at the start of every test, whether it produces output or not. Additional labels are produced as needed if interspersed output takes place, just as for `--labels:On`. Synonym for `--labels:Before`.

**--labels:Before**
A label appears at the start of every test, whether it produces output or not. Additional labels are produced as needed if interspersed output takes place, just as for `--labels:On`. Synonym for `--labels:All`.

**--labels:After**
A label appears at the end of every test, whether it produced output or not. This label includes the pass/fail status of the test in addition to its name. Additional labels are produced as needed if there is any output, just as for `--labels:On`.

### CODE REVIEW

**Comment by CharliePoole** [comment]
@nunit/core-team @nunit/engine-team @nunit/framework-team 

This could use a careful review of how it's all defined to work. I'd rather make any changes before it gets released and people are using it. In addition to whatever you come up with, I have few particular questions:

1. Should we deprecate or eliminate `--labels:All`?
2. Should the label that appears at the start of a test have a different appearance from other labels?
3. On After labels, I'm coloring only the status word (Passed, Failed, etc.) according to the result. Would it be better to color the entire line?
4. There is currently no way to get both a Before and an After label. Do we need one?

**Comment by CharliePoole** [comment]
Travis is failing because it can't get an osx agent.

**Comment by JustinRChou** [comment]
1. Doesn't seem to be a point in keeping `--labels:All`.
2. Don't quite understand this.
3. Should be fine. You do mean that it's the status color followed by the blue label?
4. I would think that should actually be the `--labels:All` option?

Should the default for labels be updated to `Before`? And is there still a need for `On`?
I guess it would not make sense for projects for lots of tests and have the default be `Before`.

**Comment by CharliePoole** [comment]
@JustinRChou You're keeping me on my toes - I should have explained some things better.

1. We can't drop it completely without breaking scripts, but I'd definitely like to phase it out.

2. Imagine you see something like this:
```
=> Name.Of.Test
Some output
```

You can't tell if that label was displayed because of the output or if it marks the start of a test. With just one test, it doesn't matter but if many tests are running in parallel and if you are trying to see how they overlap it may be important. Of course, __you__ know what option you used to run the test so you can interpret, but if the report is saved, others may not be able to tell. So there is an issue, but I'm just not sure if it's really important.

3. Exactly.

4. Yes, it's too bad that we already use All to mean something else: display label at the start of __all__ tests. This is one reason I want to phase it out: it looks like it might mean something else. And after we phase it out, I don't think we can re-use it with a different meaning without confusing people.

So basically, there is no way currently to show a label both when the test starts as well as the end. If you use Before, you see the start but you only see where it ended if there is normal output. If you use After, you can see the end but not the start. 

Should we have Both? Or wait till it's asked for?

**Comment by JustinRChou** [comment]
2. For the parallel testcase, would it produce something like
```
=> Test1(1)
=> Test2(2)
Output from Test2(2)
Output from Test1(1)
=> Test3(3)
Output from Test2(3)
```
I don't think I am good at simulating parallel tests. I might be thinking of multi-threading. 
If so, I don't see how to avoid putting the test name as part of the output again.
And when you mean saving the report, you're referring to a redirection to a log file? Or the testresult.xml?

I don't think `--labels:Both` would be critical, looks to me that `--labels:After` would be more useful. Unless 'Both' would work better for parallel tests?
But I am not a frequent user of the `label` option and parallel testing.

**Comment by JustinRChou** [comment]
Would this also be a valid solution for #25?
Looks like using `--labels:After` would provide a decent progress report.

**Comment by CharliePoole** [comment]
@JustinRChou Well, the whole point of this change is to ensure that no output appears without a proper label ahead of it (unless you turn labels off). Did you look at the tests? The ones that show two tests interacting lay out exactly the output you would see.

Yes, After can be used as a progress report of sorts just as All is currently used. But I think #25 is envisioning some kind of continued status feedback such as you often see with downloads.

**Comment by CharliePoole** [comment]
@nunit/core-team Hate to bug you guys, but this has been waiting almost three weeks for a review!

**Comment by rprouse** [comment]
Code looks good. I do worry about the number of options. It seems to me we are adding complexity and flexibility because of a vocal minority. This is also yet another change that needs to be ported across to NUnitLite. Speaking of that, I wonder if NUnitLite should be scaled back to just support simple running, exploring and filters?

**Comment by CharliePoole** [comment]
I've been thinking about scaling back NUnitLite too. It might be something I'd like to do once I get off of the framework and console.
